applications/StructuralMechanicsApplication/tests/cpp_tests/test_solid_static_sensitivity.cpp

```diff
@@ -0,0 +1,208 @@
+        KratosComponents<ConstitutiveLaw>::Get("LinearElasticPlaneStrain2DLaw"),
+        [](ModelPart* pModelPart) {
+            pModelPart->GetNode(1).Fix(DISPLACEMENT_X);
+            pModelPart->GetNode(1).Fix(DISPLACEMENT_Y);
+            pModelPart->GetNode(3).Fix(DISPLACEMENT_X);
+            pModelPart->GetNode(3).Fix(DISPLACEMENT_Y);
+        });
+    const unsigned response_node_id = 2;
+    smtsss::PrimalTestSolver solver{&primal_model_part, response_node_id};
+    const double delta = 1e-7;
+    const double response_value0 = solver.CalculateResponseValue();
+    ModelPart& adjoint_model_part =
+        CreateStructuralMechanicsAdjointTestModelPart(&primal_model_part);
+    smtsss::AdjointTestSolver adjoint_solver{&adjoint_model_part, response_node_id};
+    for (unsigned i_node : {1, 2, 3})
+    {
+        for (char dir : {'x', 'y'})
+        {
+            const double response_value1 =
+                solver.CalculateResponseValue(i_node, dir, delta);
+            const double finite_diff_sensitivity =
+                -(response_value1 - response_value0) / delta;
+            const double adjoint_sensitivity =
+                adjoint_solver.CalculateSensitivityValue(i_node, dir);
+            KRATOS_CHECK_NEAR(finite_diff_sensitivity, adjoint_sensitivity, 1e-8);
+        }
+    }
+}
+} // namespace Testing
+
+namespace
+{
+namespace smtsss
+{ // cotire unity guard
+AdjointResponseFunction::Pointer ResponseFunctionFactory(ModelPart* pModelPart, unsigned ResponseNodeId)
+{
+    Parameters params{R"({"traced_dof": "DISPLACEMENT_Y", "gradient_mode": "semi_analytic", "step_size": 1e-2})"};
+    params.AddEmptyValue("traced_node_id");
+    params["traced_node_id"].SetInt(ResponseNodeId);
+    return Kratos::make_shared<AdjointNodalDisplacementResponseFunction>(*pModelPart, params);
+}
+
+PrimalTestSolver::PrimalTestSolver(ModelPart* pPrimalModelPart, unsigned ResponseNodeId)
+    : mpPrimalModelPart(pPrimalModelPart), mResponseNodeId(ResponseNodeId){};
+
+double PrimalTestSolver::CalculateResponseValue()
+{
+    auto p_solver = CreateSolvingStrategy();
+    p_solver->Initialize();
+    p_solver->Solve();
+    auto p_response_function = ResponseFunctionFactory(mpPrimalModelPart, mResponseNodeId);
+    return p_response_function->CalculateValue(*mpPrimalModelPart);
+}
+
+double PrimalTestSolver::CalculateResponseValue(const unsigned NodeToPerturb,
+                                                const char Direction,
+                                                const double Perturbation)
+{
+    KRATOS_ERROR_IF(Direction != 'x' && Direction != 'y')
+        << "invalid direction: '" << Direction << "'";
+    KRATOS_ERROR_IF(Perturbation <= 0.) << "invalid perturbation: " << Perturbation;
+    const unsigned i_dir = (Direction == 'x') ? 0 : 1;
+    auto& r_node = mpPrimalModelPart->GetNode(NodeToPerturb);
+    r_node.GetInitialPosition()[i_dir] += Perturbation;
```

Comment:
is it enough to perturb the initial position in this case?

Reply:
Yes, I think so. The X(), Y(), Z() get recomputed during the solution.